backend/mem/location.go

```diff
@@ -0,0 +1,211 @@
+
+	list := make([]string, 0)
+	str := l.Path()
+	for _, v := range fileList {
+		if v != nil {
+			fullPath := v.Path()
+			if utils.AddTrailingSlash(path.Dir(fullPath)) == str {
+				if systemMap[fullPath] != nil {
+					existence, _ := systemMap[fullPath].Exists()
+					if existence {
+						list = append(list, v.Name())
+					}
+				}
+			}
+		}
+	}
+	return list, nil
+}
+
+/*
+ListByPrefix tags a prefix onto the current path and in a slice,
+ returns all file base names whose full paths contain that substring
+Returns empty slice if nothing found
+*/
+func (l *Location) ListByPrefix(prefix string) ([]string, error) {
+
+	list := make([]string, 0)
+	str := path.Join(l.Path(), prefix)
+	for _, v := range fileList {
+		if v != nil {
+			path := v.Path()
+			tmp := strings.Contains(path, str)
+			if tmp {
+				list = append(list, v.Name())
+
+			}
+		}
+	}
+	return list, nil
+}
+
+/*
+ListByRegex takes a regular expression and returns
+a slice of strings containing the base names of files
+found that matched the regular expression.  Returns an
+empty slice upon nothing found
+*/
+func (l *Location) ListByRegex(regex *regexp.Regexp) ([]string, error) {
+
+	list := make([]string, 0)
+	str := l.Path()
+	filesHere, _ := l.List()
+	for _, hereList := range filesHere {
+
+		potentialPath := path.Join(str, hereList)
+
+		for _, systemFileList := range fileList {
+
+			if systemFileList != nil && systemFileList.Path() == potentialPath {
+				ex, _ := systemFileList.Exists()
+				if regex.MatchString(path.Base(potentialPath)) && ex {
+					list = append(list, systemFileList.Name())
+
+				}
+
+			}
+		}
+
+	}
+
+	return list, nil
+}
+
+//Volume returns the volume of the current fs. In-Memory-Fs has no volume
+func (l *Location) Volume() string {
+	return ""
+}
+
+//Path returns the full, absolute path of the location with leading and trailing slashes
+func (l *Location) Path() string {
+
+	return l.name
+
+}
+
+//Exists reports whether or not a location exists. Creating a location does not guarantee its existence
+func (l *Location) Exists() (bool, error) {
+
+	data, _ := l.List()
+	if len(data) == 0 {
+		return false, nil
+	}
+	return true, nil
+
+}
+
+/*
+NewLocation creates a new location at the
+given relative path, which is tagged onto the current locations absolute path
+*/
+func (l *Location) NewLocation(relativePath string) (vfs.Location, error) {
+
+	str := path.Join(l.Path(), relativePath)
+	str = utils.AddTrailingSlash(path.Clean(str))
+	return &Location{
+		fileSystem: l.fileSystem,
+		name:       str,
+		exists:     false,
+	}, nil
+
+}
+
+//ChangeDir simply changes the directory of the location
+func (l *Location) ChangeDir(relativePath string) error {
+	l.name = path.Join(l.name, relativePath)
+	return nil
+
+}
+
+//FileSystem returns the type of filesystem location exists on, if it exists at all
+func (l *Location) FileSystem() vfs.FileSystem {
+
+	filePath := path.Join(path.Clean(l.Path()), l.Filename)
+	if systemMap[filePath] != nil {
```

Comment:
That's not how you check maps. Go maps also return a second variable whose boolean value will tell you if the value exists in the map.

``` go
if _, ok := mymap[badvalue]; ok {
    //my code
}
```